プラクティス「devise を使ってユーザー認証を実装する」の課題提出

nekonote-civ opened the pull request:
# 概要

プラクティス「devise を使ってユーザー認証を実装する」の課題提出

# 詳細

devise を用いたユーザー認証の実装、及び各ページの実装などを行いました。
ログイン/ログアウトなどの操作は画面上部のリンクにて行えるようにしています。

お手数おかけしますが、レビューを宜しくお願い致します。

nekonote-civ committed:
devise の導入と初期設定

nekonote-civ committed:
devise の日本語化

nekonote-civ committed:
profiles model を導入し、users の登録時に同時に登録されるようにした

nekonote-civ committed:
profiles を users に紐づけて form から登録出来るようにした

nekonote-civ committed:
ユーザーの登録後にユーザー詳細ページへリダイレクトするようにした

nekonote-civ committed:
ユーザーの一覧ページを作成し、重複するレイアウトを分割して読み込むようにした

nekonote-civ committed:
User 情報更新時に Profile も更新を行い、更新後は個別ページへリダイレクトするようにした

nekonote-civ committed:
letter_opener の導入

nekonote-civ committed:
profiles の text カラムをリネーム

nekonote-civ committed:
ルートパスを設定

nekonote-civ committed:
画面上部へユーザーのサインイン/サインアウト/編集のリンクを追加

nekonote-civ committed:
ユーザーのプロフィールページへのリンクを追加

nekonote-civ committed:
登録後のリダイレクト先をトップへ変更

nekonote-civ committed:
ページ上部のリンクへ本の一覧などを追加し、対応するCSSファイルを追加した

nekonote-civ committed:
ユーザー一覧ページへユーザーのプロフィールページへのリンクを追加

nekonote-civ committed:
各ページを日本語化し、細かい見た目を調整した

nekonote-civ committed:
rubocop の対応

nekonote-civ committed:
root の記述位置を変更

yoshitsugu commented on db/migrate/20230314083100_rename_text_column_to_introduction.rb:
```diff
@@ -0,0 +1,5 @@
+class RenameTextColumnToIntroduction < ActiveRecord::Migration[7.0]
+  def change
+    rename_column :profiles, :text, :introduction
```
同じPR内であれば影響範囲はご自身の開発環境のみですので、migrationファイルを直接変更してしまってかまいません。
実行時間の短縮にもなりますので、こういう修正系はまとめてしまいましょう。

yoshitsugu commented on db/migrate/20230312141704_create_profiles.rb:
```diff
@@ -0,0 +1,12 @@
+  def change
+    create_table :profiles do |t|
+      t.references :user, null: false, foreign_key: true
+      t.string :zipcode, limit: 7
+      t.string :address
+      t.text :text
```
validationで必須チェックをしているので `null: false` にしてしまってもいいかもしれませんね。

nekonote-civ replied:
確かに画面側だけでの validation だと問題が発生してしまいそうですね。
DB側も `null: false` にして `NOT NULL制約` を追加しました。

yoshitsugu commented on db/migrate/20230312141704_create_profiles.rb:
```diff
@@ -0,0 +1,12 @@
+class CreateProfiles < ActiveRecord::Migration[7.0]
+  def change
+    create_table :profiles do |t|
+      t.references :user, null: false, foreign_key: true
```
一人のユーザーに対して必ず一つのprofileがひもづく形になるかと思うので、 unique indexをつけてもよいかもしれないですね。

nekonote-civ replied:
`unique index` を付与し、一意のインデックスとなるように変更しました。

yoshitsugu reviewed (commented):
コメントしました。

nekonote-civ committed:
profiles テーブルを修正

user_id へ unique index の追加
別ファイルでマイグレーションしていた introduction カラムへのリネームを元ファイルで実施
バリデーションで必須項目としているカラムを not null へ変更

yoshitsugu commented on app/views/users/_user.html.erb:
```diff
@@ -0,0 +1,21 @@
+
+  <p>
+    <strong><%= Profile.human_attribute_name(:zipcode) %>: </strong>
+    <%= user.profile.zipcode %>
```
発展的な話題にはなりますが、一覧のときにN+1問題が発生しちゃってますね。対策調べてみてください〜

yoshitsugu reviewed (commented):
一点だけコメントしました。余裕あればとりくんでみてください！

nekonote-civ committed:
eager_load で N+1 を解消

nekonote-civ commented:
N+1問題修正してみましたので、ご確認お願いします！
(操作ミスの影響で force-pushed が出てしまっていますが気にしないでください🙇‍♂️)

yoshitsugu reviewed (approved):
OKです〜

nekonote-civ commented:
ありがとうございます！
